google-cloud-storage/src/main/java/com/google/cloud/storage/WriteFlushStrategy.java

```diff
@@ -199,6 +264,82 @@ public void close(@Nullable WriteObjectRequest req) {
     }
   }
 
+  public static final class DefaultBidiFlusher implements BidiFlusher {
+
+    private final BidiStreamingCallable<BidiWriteObjectRequest, BidiWriteObjectResponse> write;
+    private final RetryingDependencies deps;
+    private final ResultRetryAlgorithm<?> alg;
+    private final String bucketName;
+    private final LongConsumer sizeCallback;
+    private final Consumer<BidiWriteObjectResponse> completeCallback;
+    private final Supplier<GrpcCallContext> baseContextSupplier;
+    private volatile ApiStreamObserver<BidiWriteObjectRequest> stream;
+
+    private final BidiObserver responseObserver;
+
+    private DefaultBidiFlusher(
+        BidiStreamingCallable<BidiWriteObjectRequest, BidiWriteObjectResponse> write,
+        RetryingDependencies deps,
+        ResultRetryAlgorithm<?> alg,
+        String bucketName,
+        LongConsumer sizeCallback,
+        Consumer<BidiWriteObjectResponse> completeCallback,
+        Supplier<GrpcCallContext> baseContextSupplier) {
+      this.write = write;
+      this.deps = deps;
+      this.alg = alg;
+      this.bucketName = bucketName;
+      this.sizeCallback = sizeCallback;
+      this.completeCallback = completeCallback;
+      this.baseContextSupplier = baseContextSupplier;
+      this.responseObserver = new BidiObserver(sizeCallback, completeCallback);
+    }
+
+    public void flush(@NonNull List<BidiWriteObjectRequest> segments) {
+      ensureOpen();
+      Retrying.run(
+          deps,
+          alg,
+          () -> {
+            boolean first = true;
+            for (BidiWriteObjectRequest message : segments) {
+              message = possiblyPairDownBidiRequest(message, first);
+
+              stream.onNext(message);
+              first = false;
+            }
+            BidiWriteObjectRequest message =
+                BidiWriteObjectRequest.newBuilder().setFlush(true).setStateLookup(true).build();
+            stream.onNext(message);
```

Comment:
IIUC the main difference between bidi and non-bidi is that stream.onComplete() is not called until the Observer observes onCompleted?

Reply:
Correct, current flow calls onComplete on each flush. Bidi will only call onComplete once, at the end of the upload